Aeyb/#132 assign roles to users

tlooi opened the pull request:
**Closes:** #132 

**What was done:**
- Redesign of roles page
- Assign roles to users
- Viewing user roles

## **How to test:**
- Add and delete roles from users
- Ensure all 'tabs' work

**Screenshots of UI changes (optional):**
![image](https://github.com/UoaWDCC/aeyb-monitor/assets/108048091/457eb81f-b688-4f79-b0a7-9c8425a4ca12)
![image](https://github.com/UoaWDCC/aeyb-monitor/assets/108048091/8a6ff7d7-96fe-4b62-be49-efd61185365f)
![image](https://github.com/UoaWDCC/aeyb-monitor/assets/108048091/707bb51e-2b0f-4e12-8c6f-9286df4be16f)
![image](https://github.com/UoaWDCC/aeyb-monitor/assets/108048091/4e40907a-72d7-40e1-b297-564cde485c8e)

tlooi committed:
initial tabs

tlooi committed:
fix scrolling pages on default

tlooi committed:
add user page with roles

tlooi committed:
add check to ensure don't remove admin or default roles

tlooi committed:
add removing roles on client

tlooi committed:
prevent tab content body from overflowing

tlooi committed:
feat: add roles

tlooi committed:
lots of nested tabs :D

tlooi committed:
add mui popup

jacobatan committed:
Added new query method

jacobatan committed:
Updated remove method to do logic on DB

tlooi committed:
feat: implement permission modification on roles

Co-authored-by: jacobatan <jacob.tan229@gmail.com>

tlooi committed:
refactor: move tabs into separate component

tlooi committed:
refactor: extract out ViewRolesWindow and ViewPermissions component

tlooi committed:
refactor: moved out UserRoleRow

tlooi committed:
refactor: separate out ViewUserWindow

tlooi committed:
refactor: cleanup code imports

tlooi committed:
style: fix + button centering and add fake profile component

tlooi committed:
feat: add role button disappears if no more to add

tlooi committed:
style: prevent select of text

tlooi committed:
feat: create new roles and assign permissions

tlooi committed:
fix: color not generating properly

tlooi committed:
refactor: clean up imports

tlooi committed:
refactor: remove unused folders and removed commented lines

jacobatan committed:
Minor changes

jacobatan commented on frontend/src/pages/roles/components/ViewUserWindow.tsx:
```diff
@@ -0,0 +1,72 @@
+
+    const userContext = useUserContext();
+    async function removeRole(role: RoleDTO, userId: string) {
+        if (['Admin', 'Default'].indexOf(role.name) !== -1) {
```
Thoughts on changing this to
``` js
    if (['Admin', 'Default'].includes(role.name)) {
        return;
    }
```

Mostly because it is a lot readable. Though, we may have to take a slight performance hit, however it is so small. Thus IMO readability > the tiny bit of performance

tlooi replied:
done

jacobatan commented on frontend/src/pages/roles/components/UserRoleRow.tsx:
```diff
@@ -0,0 +1,96 @@
+    const ref = useRef(null);
+
+    function availableRolesToAdd() {
+        const rolesToAdd = roles.filter(
```
Let's break this up just for readability.

I suggest something like this:

```js
const currentRoleNames = ['Admin', 'Default', ...user.roles.map((role) => role.name)]
const rolesToAdd = roles.filter((role) => !currentRoleNames.includes(role.name));
```

tlooi replied:
done

jacobatan commented on frontend/src/pages/roles/components/UserRoleRow.tsx:
```diff
@@ -0,0 +1,96 @@
+                        return (
+                            <span
+                                onClick={() => removeRole(role, user.id)}
+                                key={`${user.name} ${role.id} ${Math.random()}`}
```
Let's remove `Math.random()`. We don't want unnecessary re-renders.

tlooi replied:
done

jacobatan commented on frontend/src/pages/roles/components/ViewPermissions.tsx:
```diff
@@ -0,0 +1,92 @@
+
+        const perms = new Set(role.permissions);
+        if (perms.has(perm)) {
+            console.log('delete');
```
Let's remove this clg

tlooi replied:
done

jacobatan commented on frontend/src/pages/roles/components/ViewRolesWindow.tsx:
```diff
@@ -0,0 +1,62 @@
+    roles: Record<string, RoleDTO>;
+    setRoles: (roles: Record<string, RoleDTO>) => void;
+}) {
+    const content = Object.keys(roles).map((id) => {
```
Let's use an implicit return here,

```
const content = Object.keys(roles).map(id => ({
    tabTitle: roles[id].name,
    tabContent: {...roles[id]},
}));

```

tlooi replied:
done

jacobatan commented on frontend/src/pages/roles/components/ViewUserWindow.tsx:
```diff
@@ -0,0 +1,72 @@
+
+    return (
+        <div className="p-4 flex flex-col gap-3 overflow-scroll w-full">
+            {content.map((data, index) => {
```
We can use an implicit return here too

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/Wowzer.tsx:
```diff
@@ -0,0 +1,26 @@
+import TabManager from '../../../utility_components/tabs/TabManager';
+
+export function Wowzer({ count = 10 + Math.random() }) {
```
Let's delete this haahahha

tlooi replied:
100% removed lel

jacobatan commented on frontend/src/pages/roles/Roles.tsx:
```diff
@@ -53,85 +50,65 @@ function Roles() {
+                            <TabManager
+                                orientation="column"
+                                content={[
+                                    { tabTitle: 'Thexboss', tabData: ['Awesome', '🥚'] },
```
LMFAO, lets get rid of this ahhahahahhaha

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/ViewPermissions.tsx:
```diff
@@ -0,0 +1,92 @@
+    savePermissions: (role: RoleDTO) => void;
+}) {
+    function checkSwitch(perm: Permission) {
+        if (['Admin', 'Default'].indexOf(role.name) !== -1) {
```
Change this to `.includes` ? Let's talk about this

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/CreateNewRole.tsx:
```diff
@@ -0,0 +1,112 @@
+                </div>
+            </div>
+            <div className="mt-4 flex flex-row-reverse">
+                <Button onClick={createNewRole} disabled={submitDisabled}>
```
Could we make this the purple aeyb colour?

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/CreateNewRole.tsx:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    useEffect(() => {
```
Let's remove this console.log

jacobatan replied:
lmfao just realised that you can't see the clg, its just the line below it

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/ViewPermissions.tsx:
```diff
@@ -0,0 +1,92 @@
+                        return (
+                            <div key={perm} className="flex justify-between items-center p-2 bg-slate-200">
+                                {perm}
+                                <Switch
```
Let's make the active colour the purple aeyb colour

tlooi replied:
Done

jacobatan commented on frontend/src/pages/roles/components/ViewRolesWindow.tsx:
```diff
@@ -0,0 +1,62 @@
+
+    return (
+        <TabManager
+            content={[{ tabTitle: 'Create New Role', tabContent: { isNotTab: true } }, ...content]}
```
I really want it to be obvious that "create a new role" is NOT a role, and that it is a button... Not sure what the clean way would be.....

tlooi replied:
Just implemented it take a look. Added an extra loader for tabs for customization

jacobatan commented:
Thoughts on adding a delete functionality to the roles?

tlooi committed:
allow customizable tabs

tlooi committed:
refactor: implement review changes

tlooi committed:
Merge branch 'AEYB/#132-Assign-roles-to-users' of https://github.com/UoaWDCC/aeyb-monitor into AEYB/#132-Assign-roles-to-users

tlooi committed:
refactor: added implicit return

tlooi committed:
style: fixed tab styling

tlooi committed:
fix: potential error by modifying JSX return to JSX element

- Prevents tabManager take in params not props to prevent accidental misuse and potential breakage of app via more hooks than prior render

tlooi committed:
revert: revert to old

jacobatan committed:
Improved code readability

tlooi committed:
feat: delete roles from database

tlooi committed:
added loaders for data

jacobatan reviewed (approved):
LGTM! One step closer to prod!
